draft-ietf-dnsop-svcb-https.md

```diff
@@ -495,59 +495,53 @@ is the effective TargetName:
-   The answers for these may or may not include CNAME pointers
-   before reaching one or more of these records.
+2. Issue a SVCB query for $QNAME.
 
-3. If an AliasMode SVCB record is returned for $SVCB_QNAME,
-   clients MUST set $ADDR_QNAME and $SVCB_QNAME to its TargetName (without
+3. If an AliasMode SVCB record is returned, set $QNAME to its TargetName (without
    additional prefixes) and loop back to step 2,
    subject to chain length limits and loop detection heuristics (see
    {{client-failures}}).
 
-4. If one or more "compatible" ({{mandatory}}) ServiceMode records are returned
-   for $SVCB_QNAME, clients SHOULD select the highest-priority compatible record.
-   This record's TargetName and SvcParams represent the preferred endpoint.  If
-   connection to this endpoint fails, the client SHOULD try to connect using
-   values from the next-highest-priority compatible record, etc. If all attempts
-   fail, clients SHOULD go to step 5 (except as noted in {{ech-client-behavior}}).
+4. If one or more "compatible" ({{mandatory}}) ServiceMode records are returned,
+   these represent the alternative endpoints.
 
-5. At this point there are no usable ServiceMode records, because
-   - there were no SVCB records found for $SVCB_QNAME, OR
-   - all records found were incompatible with this client, OR
-   - all connection attempts using ServiceMode records failed.
-
-   Accordingly, clients SHALL connect to the
-   endpoint consisting of $ADDR_QNAME, the authority endpoint's
-   port number, and no SvcParams.
-
-If all of the above connection attempts fail, clients MAY connect to the
-authority endpoint (except as noted in {{client-failures}} and
-{{ech-client-behavior}}).
+5. Otherwise, SVCB resolution has failed.
 
 This procedure does not rely on any recursive or authoritative DNS server to
 comply with this specification or have any awareness of SVCB.
 
-When selecting between AAAA and A records to use, clients may use an approach
-such as Happy Eyeballs {{!HappyEyeballsV2=RFC8305}}.
+Once SVCB resolution is completed, the client proceeds with connection
+establishment.  Clients SHOULD try higher-priority alternatives first, with
+fallback to lower-priority alternatives.  Clients MUST issue AAAA and/or A
+queries for the selected TargetName, and MAY choose between them using an
+approach such as Happy Eyeballs {{!HappyEyeballsV2=RFC8305}}.
+
+A client is called "SVCB-reliant" if it connects only to the indicated
+ServiceMode endpoints, and "SVCB-optional" otherwise.  Clients for pre-existing
+protocols SHALL implement SVCB-optional behavior (except as noted in
+{{client-failures}} and {{ech-client-behavior}}).
+
+SVCB-optional clients SHOULD issue in parallel any other DNS queries that might
+be needed for connection establishment.  SVCB-optional clients SHALL append an
+alternative endpoint consisting of the final value of $QNAME, the authority
+endpoint's port number, and no SvcParams, and MAY append the authority endpoint.
```

Comment:
I don't understand this.  Are you trying to back-fill a default value for SVCB in the case you don't need one?  Why not just say "if the endpoint does not require SVCB, then it can use A/AAAA records when SVCB is not available"?

Reply:
There are actually _two_ "default" fallbacks listed here, in a specific order and with different normativity.  The main motivation is to support using AliasMode to point to a hosting provider that doesn't use SVCB.

I've adjusted this text to improve clarity (and avoid assuming that non-SVCB connection means A/AAAA, since it could also be SRV or something else).

Reply:
As the text still says that you issue A/AAAA when you hit a ServiceMode record, I'm now more confused by this.

I'm still totally short on clarity here.  What is being appended to?  And I still don't know what "the authority endpoint" is or what it is appended to.  Are you saying that if you exhaust SVCB lookups and you are don't need the ServiceForm information, you could just pretend to have received a SVCB ServiceForm record that consists of $QNAME, the port, and no SvcParams?  Why say that at all?  What advantage does this have over letting the client fall back to non-SVCB resolution?

Reply:
> As the text still says that you issue A/AAAA when you hit a ServiceMode record, I'm now more confused by this.

Connection using a ServiceMode record always requires an address lookup for TargetName.  Connection _not_ using SVCB doesn't strictly require an address lookup ... who knows what the non-SVCB connection procedure is for an arbitrary protocol?

> What is being appended to?

The priority-ordered list of alternative endpoints.

> And I still don't know what "the authority endpoint" is or what it is appended to.

That's defined in [the definitions section](https://tools.ietf.org/html/draft-ietf-dnsop-svcb-https-02#section-1.4).

> Are you saying that if you exhaust SVCB lookups and you are don't need the ServiceForm information, you could just pretend to have received a SVCB ServiceForm record that consists of $QNAME, the port, and no SvcParams?

Yes.

> Why say that at all?

Mostly to handle the case where there is an AliasMode record but no ServiceMode record.  Adding this endpoint unconditionally for fallback is not necessary but seems like it could improve reliability (one more endpoint to try) without expanding the DNS response.

Reply:
Whereas with a precise read of the text does cover the "the case where there is an AliasMode record but no ServiceMode record" it is not obvious to a reader as shown by both Martin and I not quite getting this until reading a few times.  At a minimum we should have a parenthetical which concretely calls out this case.